src/rgw/rgw_common.h

```diff
@@ -2252,6 +2252,14 @@ struct rgw_obj {
     return index_hash_source.empty() ? key.name : index_hash_source;
   }
 
+  void maybe_update_hash_object() {
+    if (index_hash_source.empty() && key.ns == "multipart") {
+      size_t dotpos = key.name.rfind(".");  // last "." separate the part num and upload id
+      dotpos = key.name.rfind(".", dotpos - 1); // last second "." separate the upload id and rgw object name
+      index_hash_source = key.name.substr(0, dotpos);
```

Comment:
a) this string parsing logic is duplicated in `RGWMPObj::from_meta()` from services/svc_tier_rados.h, so i'd prefer to call that instead
b) please use the `#define RGW_OBJ_NS_MULTIPART "multipart"` from rgw_rados.h
c) rgw_common.h should not depend on rgw_rados.h or svc_tier_rados.h, so this logic should go directly into `RGWBucketReshard::do_reshard()`:
```diff
         rgw_obj_key key(cls_key);
         rgw_obj obj(new_bucket_info.bucket, key);
+        RGWMPObj mp;
+        if (key.ns == RGW_OBJ_NS_MULTIPART && mp.from_meta(key.name)) {
+          // place the multipart .meta object on the same shard as its head object
+          obj.index_hash_source = mp.get_key();
+        }
         int ret = store->getRados()->get_target_shard_id(new_bucket_info, obj.get_hash_object(), &target_shard_id);
```

Reply:
Thanks , those are great suggestions, I completely agree. let me modify and verify it

Reply:
done, thx, reusing the code from RGWMPObj is much more simple and reasonable